src/main/java/liquibase/ext/spanner/InsertWithSelectGenerator.java

```diff
@@ -0,0 +1,91 @@
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package liquibase.ext.spanner;
+
+import java.util.Date;
+import liquibase.database.Database;
+import liquibase.datatype.DataTypeFactory;
+import liquibase.sql.Sql;
+import liquibase.sql.UnparsedSql;
+import liquibase.sqlgenerator.SqlGeneratorChain;
+import liquibase.sqlgenerator.core.InsertGenerator;
+import liquibase.statement.DatabaseFunction;
+import liquibase.statement.core.InsertStatement;
+
+/** Generator for INSERT statements in the form 'INSERT INTO FOO (..) SELECT ...'. */
+public class InsertWithSelectGenerator extends InsertGenerator {
+
+  @Override
+  public int getPriority() {
+    // This generator should only be used manually.
+    return 0;
+  }
+
+  @Override
+  public Sql[] generateSql(InsertStatement statement, Database database,
+      SqlGeneratorChain sqlGeneratorChain) {
+    // Generate INSERT INTO (...) header.
+    StringBuilder sql = new StringBuilder();
+    sql.append("INSERT INTO ").append(database.escapeTableName(statement.getCatalogName(),
+        statement.getSchemaName(), statement.getTableName())).append(" (");
+    for (String column : statement.getColumnValues().keySet()) {
+      sql.append(database.escapeColumnName(statement.getCatalogName(), statement.getSchemaName(),
+          statement.getTableName(), column)).append(", ");
+    }
+    sql.deleteCharAt(sql.lastIndexOf(" "));
+    int lastComma = sql.lastIndexOf(",");
+    if (lastComma >= 0) {
+      sql.deleteCharAt(lastComma);
+    }
+    sql.append(")");
+
+    // Generate SELECT ... statement.
+    sql.append(" SELECT ");
+    for (String column : statement.getColumnValues().keySet()) {
+      Object newValue = statement.getColumnValues().get(column);
+      if ((newValue == null) || "NULL".equalsIgnoreCase(newValue.toString())) {
+        sql.append("NULL");
+      } else if ((newValue instanceof String)
+          && !looksLikeFunctionCall(((String) newValue), database)) {
+        sql.append(DataTypeFactory.getInstance().fromObject(newValue, database)
+            .objectToSql(newValue, database));
+      } else if (newValue instanceof Date) {
+        sql.append(database.getDateLiteral(((Date) newValue)));
+      } else if (newValue instanceof Boolean) {
+        if (((Boolean) newValue)) {
+          sql.append(DataTypeFactory.getInstance().getTrueBooleanValue(database));
+        } else {
+          sql.append(DataTypeFactory.getInstance().getFalseBooleanValue(database));
+        }
+      } else if (newValue instanceof DatabaseFunction) {
+        sql.append(database.generateDatabaseFunctionValue((DatabaseFunction) newValue));
+      } else {
+        sql.append(newValue);
+      }
+      sql.append(", ");
+    }
+
+    sql.deleteCharAt(sql.lastIndexOf(" "));
```

Comment:
So we rely on the last character always being a space, but need to check to see if it's a comma -- it doesn't make much sense to me. I assume this is taken from other Liquibase code.

Reply:
Yeah, this one is too strange. I've sanitized this generator a little bit (but yes, it was mostly re-use of code from the original generator).